docs/specs/merkle_tree.md

```diff
@@ -0,0 +1,39 @@
+# Merkle Tree
+
+A merkle tree is a data structure which is a binary tree in which the leaves are some arbitrary data and the branches each hashes of their successors. There are many ways to represent and compose merkle trees together. For that reason, we define various merkle trees, and unify them under a base interface. The primary representations of a merkle tree in coda are: an in memory merkle tree, an on disk key value database backed persistent merkle tree, and a sparse in memory merkle tree which can mask information on top of other merkle trees. Many of the choices made in the base interface of the merkle tree are specifically to enable masking while maintaining efficiency.
+
+## Base Interface
+
+The primary interactions of a merkle tree are:
+- read hash
+- read data
+- write data
+- retrieve merkle path
+
+When accessing information of the merkle tree, a merkle location is provided. A merkle location is a length encoded bitstring. Each bit represents a binary direction to take when traversing the merkle tree, starting from the first bit. If the bit is 0, traverse left, and if the bit is 1, traverse right.
+
+Reads are separated by hash and data as the return values are different. Both reads and writes are performed as bulk operations, meaning that the code that calls into these is responsible for batching as many reads and writes as it can at once. Batching of reads and writes is an important optimization for both the persistent database, and for the merkle tree masks. Writing data has the side effect of recalculating the preceeding hashes of that data.
```

Comment:
Rocksdb (and Leveldb, FWIW) have batch writes, no batch read. The mask implementation uses `Core.Hashtbl`, which has no batch operations.

We could create batch read operations, but not sure whether that would save anything, since we still have the one-by-one bottleneck of the database and hash tables.